File: HeliosEsercizio/Logger/Output/LogFile.vb
Imports Logger.Data
Imports System
Imports System.Collections.Generic
Imports System.IO
Imports System.Linq
Imports System.Text

Public Class LogFileConfiguration
End Class
Namespace Output

    Public Class LogFile
        Implements ILogOutput

        Private Const DEFAULT_FILENAME_TEMPLATE As String = "log_@Date@.log"
        Public paths As String()
        Public filenameTemplates As String()
        Public template As String
        Public application As String
        Public daysBeforeDelete As Integer

        Private _AlwaysLog As Boolean = False
        Private _ResetTime As Boolean = False
        Private _LogPreviousExceptions As Boolean = True
        Public Property AlwaysLog As Boolean Implements ILogOutput.AlwaysLog
            Get
                Return _AlwaysLog
            End Get
            Set(value As Boolean)
                _AlwaysLog = value
            End Set
        End Property

        Public Property ResetTime As Boolean Implements ILogOutput.ResetTime
            Get
                Return _ResetTime
            End Get
            Set(value As Boolean)
                _ResetTime = value
            End Set
        End Property

        Public Property LogPreviousExceptions As Boolean Implements ILogOutput.LogPreviousExceptions
            Get
                Return _LogPreviousExceptions
            End Get
            Set(value As Boolean)
                _LogPreviousExceptions = value
            End Set
        End Property


        Public Sub New(ByVal path As String, ByVal fileNameTemplate As String, ByVal templateText As String, ByVal Optional daysBeforeDelete As Integer = 0, ByVal Optional application As String = Nothing)
            Me.daysBeforeDelete = daysBeforeDelete
            Me.application = application
            Dim pathIndexes As List(Of Integer) = New List(Of Integer)()

            For Each level As Integer In [Enum].GetValues(GetType(LogLevel.Levels))
                pathIndexes.Add(level)
            Next

            paths = New String(pathIndexes.Max() + 1 - 1) {}
            filenameTemplates = New String(pathIndexes.Max() + 1 - 1) {}
            pathIndexes.ForEach(Sub(i)
                                    SetPathForLevel(CType([Enum].ToObject(GetType(LogLevel.Levels), i), LogLevel.Levels), path)
                                    SetFileNameTemplateForLevel(CType([Enum].ToObject(GetType(LogLevel.Levels), i), LogLevel.Levels), fileNameTemplate)
                                End Sub)
        End Sub

        Public Sub SetFileNameTemplateForLevel(ByVal level As LogLevel.Levels, ByVal template As String)
            filenameTemplates(CInt(level)) = If(template, DEFAULT_FILENAME_TEMPLATE)
        End Sub

        Public Sub SetPathForLevel(ByVal level As LogLevel.Levels, ByVal path As String)
            If paths.Length < CInt(level) Then
                Throw New Exception("livello di log " & level & " inesistente")
            End If

            If String.IsNullOrEmpty(path) Then
                Throw New Exception("Il path è obbligatorio")
            End If

            Try
                Dim pathExists As Boolean = Directory.Exists(path)
                If Not pathExists Then Directory.CreateDirectory(path)
            Catch ex As Exception
                Throw New Exception("Errore nell'accesso al percorso " & path, ex)
            End Try

            paths(CInt(level)) = path
        End Sub

        Public Sub LogMessage(ByVal log As Log) Implements ILogOutput.LogMessage
            Dim path As String = paths(CInt(log.IdLevel))
            Dim filename As String = If(filenameTemplates(CInt(log.IdLevel)), "")
            Dim durata As String = ""

            If log.StartTime.HasValue Then
                Dim millis As Integer?
                millis = CInt((DateTime.Now - log.StartTime.Value).TotalMilliseconds)
                durata = " (" & millis & " ms)"
            End If

            filename = filename.Replace("@Level@", [Enum].GetName(GetType(LogLevel.Levels), log.IdLevel))
            filename = filename.Replace("@Application@", application)
            Dim filenameTemplate As String = filename

            If Not filename.Contains("@Date@") Then
                filename += DateTime.Today.ToString("yyyyMMdd")
            Else
                filename = filename.Replace("@Date@", DateTime.Today.ToString("yyyyMMdd"))
            End If

            If Not filename.Contains(".") Then
                filename += ".log"
            End If

            Dim fullPath As String = path & "\" & filename

            If daysBeforeDelete > 0 Then
                Dim dir As DirectoryInfo = New DirectoryInfo(path)

                For Each file As FileInfo In dir.GetFiles(filenameTemplate.Replace("@Date@", "????????"))

                    If file.LastWriteTime < DateTime.Now.AddDays(-daysBeforeDelete) Then
                        file.Delete()
                    End If
                Next
            End If

            Dim exception As String = ""

            If log.Exception IsNot Nothing Then
                Dim innerException As String = ""
                If log.Exception.InnerException IsNot Nothing Then
                    innerException = " " & log.Exception.InnerException.Message
                End If

                exception = "- " & log.Exception.Message & innerException & " " & log.Exception.StackTrace
            End If

            Dim method As String = ""

            If Not String.IsNullOrEmpty(log.Action) Then
                method = " (" & log.Action & ")"
            End If

            Dim ipAddress As String = ""

            If Not String.IsNullOrEmpty(log.IpAddress) Then
                ipAddress = " - (" & log.IpAddress & ")"
            End If

            Dim user As String = ""

            If Not String.IsNullOrEmpty(log.Username) Then
                user = " - (" & log.Username & ")"
            End If

            Dim level As String = [Enum].GetName(GetType(LogLevel.Levels), log.IdLevel)

            Using stream As FileStream = New FileStream(fullPath, FileMode.Append, FileAccess.Write, FileShare.Write, 4096, True)

                Using sw As StreamWriter = New StreamWriter(stream)
                    sw.WriteLine(level & method & ":" & user & ipAddress & " - " & Now & durata & " - " & log.Message & exception)
                End Using
            End Using
        End Sub

    End Class
End Namespace

File: DomandaOnLine/Logger/Output/LogFile.cs
using Logger.Data;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;

namespace Logger.Output
{
	public class LogFileConfiguration
	{
		
	}
	public class LogFile : ILogOutput
	{
		private const string DEFAULT_FILENAME_TEMPLATE = "log_@Date@.log";

		public string[] paths;
		public string[] filenameTemplates;
		public string template;
		public string application;
		public int daysBeforeDelete;

		public bool AlwaysLog => false;
		public bool ResetTime => true;

		public bool LogPreviousExceptions => true;

		/// <summary>
		/// 
		/// </summary>
		/// <param name="path"></param>
		/// <param name="fileNameTemplate">
		///		Template del nome file. Possono essereutilizzati i seguenti tag:
		///		@Level@ Livello d'errore ("Debug,Info,Warning,Error,Fatal")
		///		@Application@ Nome dell'applicazione che ha generato il log
		///		@Date@ Data (in AAAAMMGG) dell'evento di log
		///		Se omesso il tag @Date@ , verrà aggiunto in coda al nome del file.
		///		Se omessa l'estenzione del file verrà messo il valore di default ".log"
		///		A.E "@Application@_@Level_@Date.txt" creerà un file del tipo "Test_Warning_20200101.txt"
		///		Se omesso, il valore di default è "log_@Date@.log"
		/// </param>
		/// <param name="templateText"></param>
		/// <param name="daysBeforeDelete">
		///		Numero di giorni in cui i file vengono mantenuti. tutti i file che sono stati creati da più giorni dalla data specificata verranno cancellati.
		///		Se impostato 0 i file non verranno mai cancellati
		///		Il valore di default è 0
		/// </param>
		/// <param name="application"></param>
		/// <example>
		///	
		/// </example>
		public LogFile(string path, string fileNameTemplate, string templateText, int daysBeforeDelete=0,string application=null){
			this.daysBeforeDelete = daysBeforeDelete;
			this.application = application;
			List<int> pathIndexes = new List<int>();
			foreach (int level in Enum.GetValues(typeof(LogLevel.Levels)))
			{
				pathIndexes.Add(level);
			}
			paths = new string[pathIndexes.Max()+1];
			filenameTemplates = new string[pathIndexes.Max()+1];
			
			pathIndexes.ForEach(i=> {
				SetPathForLevel((LogLevel.Levels)Enum.ToObject(typeof(LogLevel.Levels), i), path);
				SetFileNameTemplateForLevel((LogLevel.Levels)Enum.ToObject(typeof(LogLevel.Levels), i), fileNameTemplate);

			});
		}
		public void SetFileNameTemplateForLevel(LogLevel.Levels level, string template)
		{
			filenameTemplates[(int)level] = template?? DEFAULT_FILENAME_TEMPLATE;
		}


		public void SetPathForLevel(LogLevel.Levels level, string path)
		{
			if (paths.Length<(int)level)
			{
				throw new Exception($"livello di log {level} inesistente");
			}
			if (string.IsNullOrEmpty(path))
			{
				throw new Exception("Il path è obbligatorio");
			}
			try
			{
				bool pathExists = Directory.Exists(path);
				if (!pathExists)
					Directory.CreateDirectory(path);
			}
			catch (Exception ex)
			{
				throw new Exception($"Errore nell'accesso al percorso {path}", ex);
			}
			paths[(int)level] = path;
		}

		public void LogMessage(Log log)
		{
			string path = paths[(int)log.IdLevel];
			string filename = filenameTemplates[(int)log.IdLevel] ??"";
			string durata = "";
			if (log.StartTime.HasValue)
			{
				int? millis;
				millis = (int)(DateTime.Now- log.StartTime.Value).TotalMilliseconds;
				durata = $" ({millis} ms)";
			}
			filename = filename.Replace("@Level@", Enum.GetName(typeof(LogLevel.Levels),log.IdLevel));
			filename = filename.Replace("@Application@", application);
			string filenameTemplate = filename;
			if (!filename.Contains("@Date@"))
			{
				filename +=DateTime.Today.ToString("yyyyMMdd");
			}
			else
			{
				filename = filename.Replace("@Date@", DateTime.Today.ToString("yyyyMMdd"));
			}
			if (!filename.Contains("."))
			{
				filename += ".log";
			}
			string fullPath = $"{path}\\{filename}";

			if (daysBeforeDelete>0)
			{
				DirectoryInfo dir = new DirectoryInfo(path);
				foreach (FileInfo file in dir.GetFiles(filenameTemplate.Replace("@Date@","????????")))
				{
					if (file.LastWriteTime<DateTime.Now.AddDays(-daysBeforeDelete))
					{
						File.Delete(file.FullName);
					}
				}
			}
			string exception = "";
			if (log.Exception!=null)
			{
				exception = $" - {log.Exception.Message} {log.Exception.InnerException?.Message} {log.Exception.StackTrace}";
			}
			string method = "";
			if (!string.IsNullOrEmpty(log.Action))
			{
				method = $" ({log.Action})";
			}
			string ipAddress = "";
			if (!string.IsNullOrEmpty(log.IpAddress))
			{
				ipAddress = $" - ({log.IpAddress})";
			}
			string user = "";
			if (!string.IsNullOrEmpty(log.Username))
			{
				user = $" - ({log.Username})";
			}
			string level = $"{Enum.GetName(typeof(LogLevel.Levels), log.IdLevel)}";
			using (FileStream stream = new FileStream(fullPath, FileMode.Append, FileAccess.Write, FileShare.Write, 4096, true))
			using (StreamWriter sw = new StreamWriter(stream))
			{
				sw.WriteLineAsync($"{level}{method}:{user}{ipAddress} - {DateTime.Now}{durata} - {log.Message}{exception}");
			}
		}


	}
}
